### PR TITLE
[base] Define schema type part for better plugin support

### PR DESCRIPTION
**Pull request by rexxars**
We talked about this briefly back in August, but I thought I would take an initial stab at it;

By having a designated schema type part that plugins can implement, you are able to import schema types without explicitly defining each and every one. By having it simple by an array, you can still do operations such as filtering or mapping to remove or customize certain types, and you can also choose not to use the part at all and stick with explicit imports.

What's lacking currently is some sort of neat way to hint that an input component can be used for the type, without hard-wiring it. I'd be interested in hearing if @bjoerge has any suggestions or ideas here.

### CODE REVIEW

**Comment by bjoerge** [comment]
I can dig this!

Some considerations/requirements that I can think of:
- You should be able to handle a case where two plugins adds a type with the same name?
- You should be able to rename a type, in order to cope with the above, or in cases where it makes mores sense to call the type something else.
- A lot of plugins may export types that you don't necessarily need and you should be able to not include types from these plugins

I think your proposal covers these points nicely in that you'll have to import all schema type implementors and explicity add them to your own schema.

When it comes to components, we could also define a part that input resolvers can implement, so that you can do something along the lines of:

```js
import inputResolvers from 'all:part:@sanity/form-builder/input-resolver'

// default input resolver
export function resolveInput(type) {
  return inputResolvers.find(resolver => resolver.getInputFor(type))
}
```
In that case we should probably stub out this in the studio template much like we do with createSchema.

**Comment by bjoerge** [comment]
Side point: should't the title here be about "part", not "role"? 😆

**Comment by rexxars** [comment]
We discussed this IRL, but I thought I'd just drop a few lines here to summarize what we decided:

TL;DR: Leave it up to the developer to handle this in the way he prefers

## Two plugins adds a type with the same name, or rename a type?
```js
const resolvedTypes = schemaTypes.map(type => {
  if (type.name === 'pertEstimate' && type.fields.find(field => field.name === 'calculated')) {
    return {...type, name: 'automaticPertEstimate'}
  }

  return type
})
```

## Only including certain types?
```js
const blacklist = ['unwantedType', 'ickyType']
const filteredTypes = schemaTypes.filter(type => !blacklist.includes(type))
```

For the last point, input resolvers, I think we can start by simply defining the default input on the schema type directly, then perhaps implement a custom part for resolvers later.
